### PR TITLE
refactor(filter, inspect): adds helpers and guards to clarify code

### PR DESCRIPTION
**Pull request by vherr2**
Context
---
* Creates private guard for checking against nil/empty values
* Moves variable assignment outside of function heads when not required for differentiating pattern matching
* Adds helpers for parsing attributes and relationships

### CODE REVIEW

File: lib/ash/filter/filter.ex [hunk 1]
```diff
@@ -857,14 +857,10 @@ defimpl Inspect, for: Ash.Filter do
...
 
-  def inspect(%Ash.Filter{impossible?: impossible, not: not_filter} = filter, opts)
-      when not is_nil(not_filter) do
-    impossible =
-      if impossible do
-        "X"
-      else
-        ""
-      end
+  defguardp is_empty(val) when is_nil(val) or length(val) == 0
```
Comment (zachdaniel) [inline lib/ash/filter/filter.ex:860]: `val == [] or val == %{}` would be preferable, I think

Comment (zachdaniel) [reply lib/ash/filter/filter.ex:860]: for performance.

Comment (vherr2) [reply lib/ash/filter/filter.ex]: Does it actually matter? `length` should run in constant time right? Unless there's a minor perf difference in how it has to resolve the type first

Comment (zachdaniel) [reply lib/ash/filter/filter.ex]: I don't see how `length()` would run in constant time. pretty sure its a traversal in Elixir.

Comment (zachdaniel) [reply lib/ash/filter/filter.ex]: https://github.com/elixir-lang/elixir/issues/6374

Comment (zachdaniel) [reply lib/ash/filter/filter.ex]: Seems possible they might special case `length(thing) == 0` to `thing == []` though.

Comment (vherr2) [reply lib/ash/filter/filter.ex]: Ah yeah, I got `*_size` and `length` backwards. Yours is cleaner than what I had with `(is_map(...) and map_size(...) == 0) or length(...) == 0` though, so let's go with your proposal!

**Comment by zachdaniel** [comment]
🚀Thank you for your contribution! 🚀

**Pull request merged**